### ISSUE TITLE
Regression in private dependency handling in 1.14.3

### ISSUE THREAD

**Issue by himikof** [conan-io/conan#4970]
Hello,
I have the following issue in Conan 1.14.3 (but not in 1.14.2):

We have a somewhat strange dependency graph, where the same basic library L is depended on several times in the same package graph, sometimes private, sometimes not.
This should be equivalent to our graph:

```
A -> L
B -> L (private), A
C -> L (private), B
D -> L, C, B
```

This worked fine until 1.14.3 upgrade (one L version was picked and used where needed), when suddenly L got "binary: Skip" in the `conan info` output, and in the conan graph visualization nodes A, C and D point to the same L node which has white color (meaning no binary?), this wasn't the case before. So the L library now is suddenly missing when building D.

It seems that after #4937 Conan somehow reused wrong node for L from the graph?
I'm sorry that I am unable to provide a reproducible example, but I can do some queries on my local machine as the problem reproduces reliably on my local machine.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Yes, this seems a regression, will try to fix it asap. Thanks very much for reporting.

The part that now there is only one L is actually OK, because only one L can be linked with, and if you get to L by different paths, they should either conflict or be the same node. The issue is skipping the binary with the SKIP thing.

**Comment by memsharded** [comment]
Hi @himikof 

I am working on this. This is a very challenging thing (just changing complexity to "huge"), because the way you are using private dependencies is not really designed for it. You shouldn't have both public and private dependencies to the same dependency in the same graph. Private dependencies should be restricted to very extreme cases in which it is really necessary to deal with 2 versions of the same library, or something like that. Those versions should also be isolated and do not propagate, they cannot be public dependencies of the same graph.

Said that, I am working on a possible new way of managing this setup in https://github.com/conan-io/conan/pull/4987. If you could try running your project from that branch, that would be super useful, to have feedback before releasing it in a patch Conan 1.14.4 release.

The above comments still hold, there will be only one version of L, as it will be public downstream, not private, as it has public access through some branches of the graph.

**Comment by himikof** [comment]
It seems that #4987 fixes the immediate regression in our case, thanks!

Maybe we are using private dependencies wrong, but I think something like them is actually useful in real-world cases. I would like to describe the usage scenarios a bit, so maybe you could point me to an alternative and supported way of doing things. We are trying to limit exposure of low-level internal dependencies to high-level library consumers.

When A is depending on a C++ library B there are several options:
1. The public interface of B is used in the public interface of A.
    This is the simplest case: the dependency on B should be propagated to all consumers of A. 
    If B is a shared library, it should be copied on import.
    This is the behavior of simple Conan dependencies, as I understand them.

2. B is a shared library and it is not visible from the interface of A, and A is a shared library/executable.
    In this case B should not be propagated to the consumers of A, because it would expose them to a dependency which they should not be able to use without requiring it directly. It should be still copied on import.
    For this case we are currently using private dependencies in Conan, maybe we were lucky that it worked?

3. B is a shared library and it is not visible from the interface of A, and A is a static/header-only library.
     In this case the best result would be propagating only link-time dependencies (because they are not stored in the static libraries), but not include directories.
     I don't know if this is possible in Conan.

4. B is a static/header-only library and is not visible from the interface of A, and A is a shared library/executable.
    In this case binary consumers of A should never have to see anything about B, it is completely hidden from them. So we should be able to use a build time dependency.

5. B is a static library and is not visible from the interface of A, and A is a static/header-only library.
    In this case the best result would be propagating only link-time dependencies (because they are not stored in the static libraries), but not include directories.

6. B is a header-only library and is not visible from the interface of A, and A is a static/header-only library.
    In this case no dependency propagation is expected, so a private dependency should be enough.

Now that I have written all the cases, I think there are actually 3 separate aspects of a dependency:
1. Whether it should be used for binary packages, or only for source builds
2. Which parts of cpp_info should be propagated to consumers (likely to be better expressed in terms of api/implemetation dependency kinds and header-only/static/shared/executable target types).
3. Whether the dependency ABI affects the package ABI (package_id propagation)

Currently the controls seem too coarse-grained, and often affects several of these aspects at once in a non-trivial manner (for example, switching requires->build_requires affects all 3, I think). Some parts are fine-tunable manually (for example, number 3 can be customized by overriding `package_id()` and modifying `self.info.requires`), but I see no way to customize cpp_info propagation, and doing it manually for every dependency edge will be very error-prone. It would be much better if these aspects could be independently controlled in a concise way for every dependency.
I also have just found #4038, it seems very much related to what we are trying to achieve with private dependencies.

**Comment by memsharded** [comment]
> B is a shared library and it is not visible from the interface of A, and A is a shared library/executable.
In this case B should not be propagated to the consumers of A, because it would expose them to a dependency which they should not be able to use without requiring it directly. It should be still copied on import.
For this case we are currently using private dependencies in Conan, maybe we were lucky that it worked?

Yes, in theory, conan ``private`` dependencies are fully private. When the binary of A exists, the binaries of its transitive dependencies are not even downloaded from the server. It would make more sense for an static library B which is embedded by an executable or shared library A, if an only if anything of B escapes the public API of A or affects its ABI. But even in that case, it is not very recommended.

You are totally right in your "aspects" classification, we have been reviewing these concepts a couple of times in the last months to plan for improving the graph model, and they are basically the same: "build", "propagate", "affect binary". However this will be a complex thing, and we will need to keep thinking and working on it, seeing how it can be improved without breaking everything (which is proving extraordinarily challenging). 

Thanks very much for your feedback, it is very useful!

**Comment by memsharded** [comment]
P.S: Adding this issue to the "Graph & requirements" project, where all the graph-related improvements are gathered.

**Comment by memsharded** [comment]
This should be closed by https://github.com/conan-io/conan/pull/4987, and will be released in Conan 1.14.4 soon. Please check when it is released and re-open or report if necessary. Thanks!
